### ISSUE TITLE
Exposing notifications, subscribe, unsubscribe

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#153]
Currently the WoT Scripting spec uses a minimal definition of `Observable`:
```javascript
interface Observable {
  Subscription subscribe(EventHandler handler,
                         optional ErrorHandler errorHandler,
                         optional OnComplete onComplete);
};
interface Subscription {
  void unsubscribe();
  readonly attribute boolean closed;
};
callback EventHandler = void (any value);
callback ErrorHandler = void (Error error);
callback OnComplete = void ();
```

It is used by the [`wot.discover()`]() method and it is implemented by [`ConsumedThing`](https://w3c.github.io/wot-scripting-api/#the-consumedthing-interface) for observing TD changes,  [`ThingProperty`](https://w3c.github.io/wot-scripting-api/#the-thingproperty-interface) and [`ThingEvent`](https://w3c.github.io/wot-scripting-api/#the-thingevent-interface) for observing value changes.

However, verbal feedback on TPAC 2018 discouraged the use of Observables. 

Instead, we should look into applying the [TAG design document](https://w3ctag.github.io/design-principles/#event-design) that encourages using [Event](https://dom.spec.whatwg.org/#event), [EventTarget](https://dom.spec.whatwg.org/#interface-eventtarget), [AbortController](https://dom.spec.whatwg.org/#interface-abortcontroller), and [AbortSignal](https://dom.spec.whatwg.org/#abortsignal).

There are several options forward (expanded in next comments).

**Comment by zolkis** [comment]
One option is to continue using a WoT-owned definition of subscriber interface that uses callbacks either as arguments or properties, for instance the current `Observable`, or a variant like the following:

```javascript
interface Subscription {  // subscriber methods return this object
  attribute NotificationHandler? onnext;
  attribute ErrorHandler? onerror;
  attribute OnComplete? oncomplete;
  void abort(optional any cancellationData);
};
callback NotificationHandler = void (any value);
callback ErrorHandler = void (Error error);
callback OnComplete = void ();
```

The modified interfaces that use this:

```javascript
partial interface WoT {
  Subscription discover(optional ThingFilter filter);
};
partial interface ConsumedThing {
  Subscription observe();
};
partial interface ThingProperty {
  Subscription observe();
}
```

The difference from `Observable` is that the `subscribe()` method definition is "moved out" into the interfaces that use it. I think `Observable` is more usable because of simpler declaration and more straightforward usage in code.

Example code for discovery:
```javascript
let discoveryFilter = {
  method: "directory",
  url: "http://directory.wotservice.org"
};
let subscription = wot.discover(discoveryFilter);
subscription.onnext = td => {
    console.log("Found Thing " + td.name);
    // fetch the TD and create a ConsumedThing
    let thing = wot.consume(td);
  };
subscription.onerror = error => { 
  console.log("Discovery finished because an error: " + error.message); 
};
subscription.oncomplete = () => { console.log("Discovery finished successfully");};
setTimeout( () => {
    subscription.abort();
    console.log("Discovery timeout");
  },
  5000);
```

**Comment by zolkis** [comment]
Another option is to use DOM Events, EventTarget etc.

```javascript
interface Subscription: EventTarget { // subscribe methods return this object
  attribute EventHandler? onnext;  // NotificationEvent
  attribute EventHandler? onerror;  // SubscriptionErrorEvent
  attribute EventHandler? oncomplete;  // Event
};
interface WoTAbortController: AbortController {
  void abort(optional any cancellationData);  // to be double-checked if needed
};
interface NotificationEvent: Event { // received by listeners
  readonly attribute any value;
};
interface SubscriptionErrorEvent: Event { // received by listeners
  readonly attribute Error error;
};
```
The modified interfaces that use this are similar as above:
```javascript
partial interface WoT {
  Subscription discover(optional ThingFilter filter);
};
partial interface ConsumedThing {
  Subscription observe();
};
partial interface ThingProperty {
  Subscription observe();
};
```
The discovery example can look the same on subscribe-only, but when abort() is needed as well, then quite different:
```javascript
let discoveryFilter = {
  method: "directory",
  url: "http://directory.wotservice.org"
};
let aborter = new AbortController();
let signal = aborter.signal;
let subscription = wot.discover(discoveryFilter);
subscription.addEventListener("next", td => {
    console.log("Found Thing " + td.name);
    // fetch the TD and create a ConsumedThing
    let thing = wot.consume(td);
  }, { signal });
subscription.onerror = error => { 
  console.log("Discovery finished because an error: " + error.message); 
};
subscription.oncomplete = () => { console.log("Discovery finished successfully");};
setTimeout( () => {
    aborter.abort();
    console.log("Discovery timeout");
  },
  5000);
```

**Comment by zolkis** [comment]
Another option is to use a WoT-specific Observable:

```javascript
interface Observer {
  Promise<void> observe(ObserverHandler handler, ObserveOptions options);  
  // resolve when complete, reject on error, calls handler (with all samples since last call?)
};
dictionary ObserveOptions {
  AbortSignal signal;
  // could include sampling and reporting frequency, samples buffer size, policies etc
};
callback ObserverHandler = void (any value);  // a handler algorithm to be defined for each app
ConsumedThing includes Observer;
ThingProperty includes Observer;
ThingEvent includes Observer;
partial interface WoT {
  Promise<void> discover(ObserverHandler handler, optional DiscoveryOptions options);
};
dictionary DiscoveryOptions: ObserveOptions {
  ThingFilter filter; 
};
interface WoTAbortController: AbortController {
  void abort(optional any cancellationData);
};
```

**Comment by draggett** [comment]
My preference is to start simple and extend as and when we see customer pull in respect to advanced features.  Right now my implementation for the [arena-webhub](https://github.com/draggett/arena-webhub) client supports the following interface for individual properties and events:

``` JavaScript
subscribe(observer)
unsubscribe(observer)
```

where observer is a function that is passed a single argument for the corresponding updated property value or the event's data.  This actually is layered on top of functions exposed by the platform abstraction layer, that implements the different patterns of protocol usage for each platform, e.g. Things Gateway, ThingWeb and Arena.

The platform abstraction layer is responsible for establishing and restoring the underlying connections (e.g. over Web sockets) when appropriate, e.g. when a browser tab becomes visible after a long time when it is hidden.

Note that I see a need for for a class of system events that signal when connectivity is lost and established. One example is where a thing is exposed by a device behind a firewall and re-exposed by a proxy in the cloud. Applications that consume a thing exposed by the proxy will want to know whether the proxy is currently connected or not to the corresponding IoT device, given that the device itself may be offline or powered down.

Another example is where the bearer token for access to a thing is revoked or expires.  The server then closes the currently open Web Socket connections used by the client for listening for events.  The platform adaptation layer in the client then automatically attempts to re-establish the connection, and fails with a security error. This is signalled to the client application by an event that allows the application to take the appropriate action, e.g. sending an SMS to the owner to login to refresh the token.

I am thus wondering about how to reserve part of the event name space for such system events? One way is to provide a different interface for listening for such events.  You might think we could simply add the subscribe and unsubscribe methods to the thing itself, but we may in future want to allow that as a means to listen to all events on a thing. We could perhaps use _subscribeSystemEvent_ rather than _subscribe_ or we could reserve event names starting with a particular prefix. What do you think?

**Comment by zolkis** [comment]
> subscribe(observer)
unsubscribe(observer)

Right, this is somewhat similar to the Subscriber/Observer interface, plus we have _options_ dictionary (that is also used for adapting to the AbortController/AbortSignal pattern recently introduced to the Web Platform).

> Note that I see a need for for a class of system events that signal when connectivity is lost and established. 

I assume this (as other provisioning or business logic) could actually be part of the TD.

If we say we want to standardize a common subset of provisioning or connectivity handling then I guess we could pull those definitions into a TD from a well established TD-modules repository. So instead of standardizing on platform level, we standardize some descriptions and then scripts will follow that. We could avoid defining multiple namespaces then (since the TD would define it).

> I am thus wondering about how to reserve part of the event name space for such system events?

We have been cautious using software events for modeling WoT Events, because pulling in full DOM Events support would have been a bit too thick for constrained environments. For the moment WoT Events are modeled with the ThingEvent interface that implements Observables, and in the future we might use a Subscriber or Observer pattern. As a result, we could use programmatic events (DOM or Node.js) entirely to expose platform-related notifications.

Anyway, using a different subscription primitive per event namespace (WoT system/platform, WoT Events) is one solution.

However, the TAG would likely contest such design choices (as would likely do with a simple subscribe()/unsubscribe() API, by saying we should use existing mechanisms such as EventTarget or various Observers). In this issue we try to figure out how to best use the https://w3ctag.github.io/design-principles/ for our WoT use cases.

**Comment by zolkis** [comment]
@draggett : I agree with 1., tend to agree with 3., and 2. needs more discussion :).

**Comment by zolkis** [comment]
> EventTarget comes with a model of event bubbling, capture and default actions that I think would be inappropriate for things as simple objects, regardless of whether the platform is on a constrained environment. I would be happy to be part of a discussion with the TAG on that.

It would be nice if a minimal Event and EventTarget interface was defined (e.g. as SimpleEvent and SimpleEventTarget) to be able to use them in constrained environments and DOM Events would extend these.

Similarly, it would be nice to have a simple Observer and [Stream](https://streams.spec.whatwg.org/) pattern defined and current observers and streams would subclass that.

**Comment by draggett** [comment]
In respect to "2. system events above",  we could as you suggest start with these explicitly as part of the TD, and later make them implicit when we have amassed sufficient developer experience. However, I would want to see us defining some standard names for the most common events to avoid unnecessary fragmentation. We can start to create a collection of system events, e.g. loss of connectivity, expired credentials, updates to thing metadata and so forth.

**Comment by zolkis** [comment]
> we could as you suggest start with these explicitly as part of the TD

I think it makes sense to standardize the vocabulary for these system events in the TD task force.
@sebastiankb  ^^

**Comment by zolkis** [comment]
One of the alternatives that plays adheres to the TAG design guidelines and is intuitive to use in WoT code is an aligned variant of `Subscription` also defined [above](https://github.com/w3c/wot-scripting-api/issues/153#issuecomment-435386521):
```javascript
interface Subscription: EventTarget { 
  attribute EventHandler? onnext;  // ThingEvent
  attribute EventHandler? onerror; // ErrorEvent
  readonly attribute boolean observing;  // false: observation has ended/completed
  void start();
  void stop();
};
interface ThingEvent: Event {
  readonly attribute any data;
};
interface ErrorEvent: Event {
  readonly attribute Error error;
};
partial interface WoT {
  Subscription discover(optional ThingFilter filter);  // onnext event .data is of type ThingDescription
};
partial interface ConsumedThing {
  Subscription observe();  // onnext event .data is of type ThingDescription
};
partial interface ThingProperty {  // same for ThingProperty and ThingEvent
  Subscription observe();  // simple event; will update the hosting object
};
partial interface ThingEvent {  // same for ThingProperty and ThingEvent
  Subscription observe();  //  onnext event .data can be of any value, defined by the TD
};
```
Well, there is the extra step of `subscription.start()`, but it's needed since attaching event listeners cannot have side effects (i.e. we can't say that attaching the first listener will make a request to observe). A similar design has also been used in the new [Web NFC spec](https://w3c.github.io/web-nfc/#the-nfcreader-and-nfcwriter-objects).

Another alternative would be a variant of `Observer`: 
```javascript
interface ThingObserver: EventTarget {
  void observe();  // options could be added for sampling/reporting frequency etc. 
  void unobserve();  // abort ongoing observation
  readonly attribute boolean observing;  // false: observation has ended/completed
  attribute EventHandler onnext;
  attribute EventHandler onerror;
};
interface ObserveEvent: Event {
  readonly attribute any data;
};
ConsumedThing includes Observer;
ThingProperty includes Observer;
ThingEvent includes Observer;
partial interface WOT {
  ThingObserver discover(optional DiscoveryOptions options);
};
```
Using this approach will directly add events, property and methods to `ConsumedThing`, `ThingProperty` and `ThingEvent`, which may increase the likelihood of name clashing with TD defined properties since now we have the `onnext`, `onerror` and `observing` names that cannot be used in a TD definition.

**Comment by zolkis** [comment]
Fixed in #165
